plugins/woocommerce/src/Internal/Utilities/Types.php

```diff
@@ -0,0 +1,69 @@
+<?php
+
+namespace Automattic\WooCommerce\Internal\Utilities;
+
+use InvalidArgumentException;
+
+/**
+ * Utilities to help ensure type safety.
+ */
+class Types {
+	public const UNABLE_TO_ENSURE_INSTANCE_IS_OF_EXPECTED_TYPE = 'unable-to-ensure-instance-is-of-expected-type';
+
+	/**
+	 * Checks if $thing is an instance of $desired_type.
+	 *
+	 * If the check succeeds, $thing will be returned without further modification. If the check fails, then either
+	 * an exception will be thrown or, if an $on_failure callback was supplied, it will be invoked to either generate
+	 * an appropriate return value or to throw a more specific exception.
+	 *
+	 * Please note that the failure handler will be passed two arguments:
+	 *
+	 *     $on_failure( $object, $desired_type )
+	 *
+	 * @since 9.1.0
+	 * @throws InvalidArgumentException If $object does not match $desired_type, and an $on_failure callback was not supplied.
+	 *
+	 * @param mixed     $thing        The value or reference to be assessed.
+	 * @param string    $desired_type What we expect the return type to be, if it is not a WP_Error.
+	 * @param ?callable $on_failure   If provided, and if evaluation fails, this will be invoked to generate a return value.
+	 *
+	 * @return mixed
+	 */
+	public static function ensure_instance_of( $thing, string $desired_type, callable $on_failure = null ) {
+		// If everything looks good, return early.
+		if ( $thing instanceof $desired_type ) {
+			return $thing;
+		}
+
+		// Summarize the error for use in logging and in case we have to throw an exception.
+		$summary = sprintf(
+			/* translators: %1$s: name of the expected type */
+			__( 'Unable to ensure that an object was of expected type %1$s. This is not necessarily a problem within WooCommerce, and may indicate there is a problem with code from another plugin that has hooked into one or more WooCommerce filters (please refer to the backtrace).', 'woocommerce' ),
```

Comment:
This probably doesn't need to be translated (see [Logging Best Practices](https://github.com/woocommerce/woocommerce/blob/d2bde1368af18dc3489dcca96086f2986835ee1f/docs/extension-development/logging.md#best-practices)). It also seems pretty verbose to me. Could it maybe be simplified to something like "Object was not the expected type %1$s. Checking the backtrace may reveal where the object was changed."

Reply:
Yep, we can whittle it down. 

My hope was that giving a more complete steer ('this may be caused by another plugin'), and making it available in the site operator's language, we might ultimately help those folks to self-diagnose and thereby reduce support requests, but perhaps that's a bit naive (and I realize our logging system doesn't support error codes, so sticking to English is probably best, as covered in those best practices).

Will update :+1: